src/meta/src/barrier/schedule.rs

```
@@ -340,21 +339,62 @@ impl BarrierScheduler {
-        let scheduled = match queue.queue.pop_front() {
-            Some(mut scheduled) => {
+    pub(super) fn set_min_interval(&mut self, min_interval: Duration) {
+        let set_new_interval = match &self.min_interval {
+            None => true,
+            Some(prev_min_interval) => min_interval != prev_min_interval.period(),
+        };
+        if set_new_interval {
+            let mut min_interval = tokio::time::interval(min_interval);
+            min_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+            self.min_interval = Some(min_interval);
+        }
+    }
+
+    pub(super) async fn next_barrier(&mut self) -> Scheduled {
+        let scheduled = select! {
+            biased;
+            mut scheduled = async {
+                loop {
+                    let mut rx = self.inner.changed_tx.subscribe();
+                    {
+                        let mut queue = self.inner.queue.lock();
+                        if let Some(scheduled) = queue.queue.pop_front() {
+                            break scheduled
+                        }
+                    }
+                    rx.changed().await.unwrap();
+                }
+            } => {
+                if let Some(min_interval) = &mut self.min_interval {
+                    min_interval.reset();
+                }
+                let checkpoint = self.try_get_checkpoint();
                 scheduled.checkpoint = scheduled.checkpoint || checkpoint;
                 scheduled
-            }
-            None => {
-                // If no command scheduled, create a periodic barrier by default.
+            },
+            _ = async {
+                match &mut self.min_interval {
+                    Some(min_interval) => min_interval.tick().await,
+                    None => {
+                        warn!("min interval not set and pending");
+                        pending::<_>().await
+                    }
```

I think we should panic if the interval is not set.